From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TP2 BMC <tp2bmc@local>
Date: Thu, 02 Jul 2026 03:00:00 +0200
Subject: [PATCH 6/6] tpi-smi-mux: release I2C pins before GPIO request

Selecting smi_mode on the I2C controller must release the default TWI
pinctrl ownership. The Realtek SMI driver then requests PE12/PE13 through
gpiolib only for the duration of one SMI transaction and releases them
before the arbiter restores the default TWI state.

Also use pm_runtime_suspend() instead of pm_runtime_put_sync() so the
arbiter does not underflow the mv64xxx runtime PM usage count.

Signed-off-by: TP2 BMC <tp2bmc@local>
---
 drivers/i2c/busses/tpi-i2c-smi-arbiter.c |  9 +++------
 drivers/net/dsa/realtek/realtek-smi.c    | 16 +++++++++++++---
 2 files changed, 16 insertions(+), 9 deletions(-)

diff --git a/drivers/i2c/busses/tpi-i2c-smi-arbiter.c b/drivers/i2c/busses/tpi-i2c-smi-arbiter.c
index c3d4e5f6a7b8..d4e5f6a7b8c9 100644
--- a/drivers/i2c/busses/tpi-i2c-smi-arbiter.c
+++ b/drivers/i2c/busses/tpi-i2c-smi-arbiter.c
@@ -46,10 +46,8 @@ int tpi_smi_arb_begin(struct device *arb_dev)
 	mutex_lock(&arb->lock);
 	i2c_lock_bus(arb->adap, I2C_LOCK_ROOT_ADAPTER);
 
-	if (pm_runtime_enabled(arb->ctl_dev))
-		pm_runtime_put_sync(arb->ctl_dev);
-
-	pinctrl_pm_select_idle_state(arb->ctl_dev);
+	if (pm_runtime_enabled(arb->ctl_dev))
+		pm_runtime_suspend(arb->ctl_dev);
 
 	ret = pinctrl_select_state(arb->pctl, arb->st_smi);
 	if (ret) {
@@ -81,9 +79,6 @@ void tpi_smi_arb_end(struct device *arb_dev)
 
 	pinctrl_pm_select_default_state(arb->ctl_dev);
 
-	if (pm_runtime_enabled(arb->ctl_dev))
-		pm_runtime_get_sync(arb->ctl_dev);
-
 	i2c_unlock_bus(arb->adap, I2C_LOCK_ROOT_ADAPTER);
 	mutex_unlock(&arb->lock);
 }
diff --git a/drivers/net/dsa/realtek/realtek-smi.c b/drivers/net/dsa/realtek/realtek-smi.c
index cccccccc..dddddddd 100644
--- a/drivers/net/dsa/realtek/realtek-smi.c
+++ b/drivers/net/dsa/realtek/realtek-smi.c
@@ -57,11 +57,11 @@ static int realtek_smi_ensure_gpios(struct realtek_priv *priv)
 	if (priv->mdc && priv->mdio)
 		return 0;
 
-	priv->mdc = devm_gpiod_get_optional(priv->dev, "mdc", GPIOD_OUT_LOW);
+	priv->mdc = gpiod_get_optional(priv->dev, "mdc", GPIOD_OUT_LOW);
 	if (IS_ERR(priv->mdc))
 		return PTR_ERR(priv->mdc);
 
-	priv->mdio = devm_gpiod_get_optional(priv->dev, "mdio", GPIOD_OUT_LOW);
+	priv->mdio = gpiod_get_optional(priv->dev, "mdio", GPIOD_OUT_LOW);
 	if (IS_ERR(priv->mdio))
 		return PTR_ERR(priv->mdio);
 
@@ -94,8 +94,18 @@ static int realtek_smi_bus_begin(struct realtek_priv *priv)
 
 static void realtek_smi_bus_end(struct realtek_priv *priv)
 {
-	if (priv->smi_arb)
+	if (priv->smi_arb) {
+		if (priv->mdio) {
+			gpiod_put(priv->mdio);
+			priv->mdio = NULL;
+		}
+		if (priv->mdc) {
+			gpiod_put(priv->mdc);
+			priv->mdc = NULL;
+		}
+
 		tpi_smi_arb_end(priv->smi_arb);
+	}
 }
 
 static void realtek_smi_start(struct realtek_priv *priv)
